### DISCUSSION TITLE
v7: How to use context in Route.LoaderArgs

### DISCUSSION THREAD

**Discussion by nbudin** [remix-run/react-router#12314] [Q&A] [answered]
Hi community!  I'm attempting to use the new type-safe routing features in the react-router v7 prereleases.  I'd like to use the `context` argument in my loaders to pass an instance of an Apollo client that loaders can share, to avoid redundant queries in nested routes.

I haven't been able to figure out a way to augment the type of the `context` field in `Route.LoaderArgs`.  It seems like this ambient type declaration file should work:

```typescript
import { ApolloClient, NormalizedCacheObject } from '@apollo/client';

declare module 'react-router' {
  interface AppLoadContext {
    client: ApolloClient<NormalizedCacheObject>;
  }
}
```

And indeed, when I inspect the type of the `context` field in a loader, it seems to be the `AppLoadContext` type from the `react-router` module.  But the declaration above seems to have no effect.  I've already verified that the ambient type definition file is in fact being loaded and has an effect in the route module by defining a global type and attempting to use it, which works fine.

Has anyone managed to make this work?  If so, I'd love to hear how you did it.  Thanks!

**Comment by sergiodxa** [comment, marked as answer]
I had this issue on `7.0.0-pre.5`, but if you upgrade to `7.0.0-pre.6` it works again, seems they fixed it somehow.

**Reply by nbudin** [reply]
Awesome, thanks Sergio!  I just tried it and it worked for me.

**Comment by jarrodpayne** [comment]
I understand this issue is about the AppLoadContext type, but can someone share how do you actually augment the context object at runtime? I'm struggling to find it without dropping down to a custom server.

**Reply by gustavopch** [reply]
On RR7, you should just need to add this to a `.ts` or `.d.ts` in your project:
```ts
declare module 'react-router' {
  interface AppLoadContext {
    something: string
  }
}
```
If you're using `"moduleResolution": "nodenext"`, you'll also need this patch:
```diff
diff --git a/node_modules/react-router/package.json b/node_modules/react-router/package.json
index bdaf663..a3eb8b7 100644
--- a/node_modules/react-router/package.json
+++ b/node_modules/react-router/package.json
@@ -24,7 +24,7 @@
   "exports": {
     ".": {
       "node": {
-        "types": "./dist/development/index.d.ts",
+        "types": "./dist/development/index.d.mts",
         "module-sync": "./dist/development/index.mjs",
         "default": "./dist/development/index.js"
       },
```
If none of this work for you, then there's something else going on in your project and I'll suggest you create a repro with https://remix.new or https://reactrouter.com/new.

**Reply by jarrodpayne** [reply]
@gustavopch Thanks for the suggestion. I'm able to get the types to work correctly using a `.d.ts` file. I'm just unable to find where I can set the context values at runtime (other than when using a custom server).

**Reply by gustavopch** [reply]
Ah, I understand now. I only have experience using Express. I think you're basically supposed to use a custom server for anything other than very trivial demos.

**Reply by ergen35** [reply]
> @gustavopch Thanks for the suggestion. I'm able to get the types to work correctly using a `.d.ts` file. I'm just unable to find where I can set the context values at runtime (other than when using a custom server).

Just a suggestion. You could make react-router reveals it **entry.server.tsx** file (by running **npx react-router reveal**)
then you can set it into the **handleRequest** function using the **loadContext** parameter

**Reply by sergiodxa** [reply]
The `handleRequest` in entry.server runs **after** route loaders, and only for SSR. Changing the loadContext there won't really do nothing for the rest of your app. It can be used to augment the type but not set things there.
